connection.go

```diff
@@ -250,41 +261,56 @@ func logBadQueryState(log *logger.DBSQLLogger, opStatus *cli_service.TGetOperati
 func (c *conn) executeStatement(ctx context.Context, query string, args []driver.NamedValue) (*cli_service.TExecuteStatementResp, error) {
 	corrId := driverctx.CorrelationIdFromContext(ctx)
 	log := logger.WithContext(c.id, corrId, "")
-	sentinel := sentinel.Sentinel{
-		OnDoneFn: func(statusResp any) (any, error) {
-			req := cli_service.TExecuteStatementReq{
-				SessionHandle: c.session.SessionHandle,
-				Statement:     query,
-				RunAsync:      c.cfg.RunAsync,
-				QueryTimeout:  int64(c.cfg.QueryTimeout / time.Second),
-				// this is specific for databricks. It shortcuts server roundtrips
-				GetDirectResults: &cli_service.TSparkGetDirectResults{
-					MaxRows: int64(c.cfg.MaxRows),
-				},
-				// CanReadArrowResult_: &t,
-				// CanDecompressLZ4Result_: &f,
-				// CanDownloadResult_: &t,
-			}
-			ctx = driverctx.NewContextWithConnId(ctx, c.id)
-			resp, err := c.client.ExecuteStatement(ctx, &req)
-			return resp, wrapErr(err, "failed to execute statement")
-		},
-		OnCancelFn: func() (any, error) {
-			log.Warn().Msg("databricks: execute statement canceled while creation operation")
-			return nil, nil
+
+	req := cli_service.TExecuteStatementReq{
+		SessionHandle: c.session.SessionHandle,
+		Statement:     query,
+		RunAsync:      c.cfg.RunAsync,
+		QueryTimeout:  int64(c.cfg.QueryTimeout / time.Second),
+		// this is specific for databricks. It shortcuts server round trips
+		GetDirectResults: &cli_service.TSparkGetDirectResults{
+			MaxRows: int64(c.cfg.MaxRows),
 		},
+		// CanReadArrowResult_: &t,
+		// CanDecompressLZ4Result_: &f,
+		// CanDownloadResult_: &t,
 	}
-	_, res, err := sentinel.Watch(ctx, c.cfg.PollInterval, c.cfg.QueryTimeout)
-	if err != nil {
-		return nil, err
+
+	ctx = driverctx.NewContextWithConnId(ctx, c.id)
+	resp, err := c.client.ExecuteStatement(ctx, &req)
+
+	var shouldCancel = func(resp *cli_service.TExecuteStatementResp) bool {
+		if resp == nil {
+			return false
+		}
+		hasHandle := resp.OperationHandle != nil
+		isOpen := resp.DirectResults != nil && resp.DirectResults.CloseOperation == nil
+		return hasHandle && isOpen
 	}
 
-	exStmtResp, ok := res.(*cli_service.TExecuteStatementResp)
-	if !ok {
-		return exStmtResp, errors.New("databricks: invalid execute statement response")
+	select {
+	default:
+	case <-ctx.Done():
```

Comment:
Thanks for explaining this logic to me offline. Would be nice to revisit sentinel in the future to perhaps have it work for non-polling operations.

Reply:
yeap. But maybe work for post GA.